Proposals/20200511-HVPA-design.md

```diff
@@ -0,0 +1,211 @@
+# HVPA-EP-0001: HVPA Redesign
+
+## Summary
+This proposal outlines a plan to enable HVPA to do more efficient scaling horizontally and vertically.
+
+## Motivation
+Currently, HVPA scales by applying user provided weights to HPA and VPA recommendations. However, if 100% weight is provided to VPA, then horizontal scaling will never happen.
+The motivation behind this proposal is to enable users to provide the conditions, and the type of scaling for their applications under those conditions.
+
+### Goals
+
+1. HVPA should be able to scale an application as per the specifications provided by the users
+
+## Proposal
+
+Users will provide buckets of resources each of which would include minimum and maximum CPU, memory and replicas respectively. A configuration is said to belong to a bucket if total resources allocated to all its replcas lie in that bucket. For eg. if there are 3 replicas with 4 cores each, i.e., total 12 cores, then this configuration could belong to a bucket with:
+```
+minimum CPU: 3
+maximum CPU: 6
+minimum memory: 2.5G
+maximum memory: 7G
+minimum replicas: 2
+maximum replicas: 5
+```
+For this bucket mimimum and maximum total CPU are 6 (= 3 * 2) and 30 (= 6 * 5) respectively, and same for the memory would be 5G and 35G respectively.
+So a configuration with 3 replicas with 6GB memory each would also lie in the same bucket.
+
+HVPA controller would calculate the total memory and CPU recommended by the recommenders (eg. HPA or VPA) by multiplying pod resources and the number of replicas, and would find the bucket for the calculated total resources. HVPA controller would then divide the bucket linearly to calculate the actual `replicas` and `resources` to which it will scale the application such that the total allocated resources to all its replicas is same as recommeded by the recommenders.
+
+### User Stories (optional)
+
+#### Story 1
+A user wants to scale both horizontally and vertically with folloing constraints:
+1. minimum replicas: 1, maximum replicas: 8
+1. minimum memory: 500M, maximum memory: 25G
+1. mimimum CPU: 0m, maximum CPU: 24000m
+
+The above requirement can be denoted as a single bucket.
+HVPA controller would divide the "CPU sub-bucket" and the "memory sub-bucket" linearly among the replica counts such that:
```

Comment:
do you mean that we divide, for eg., memory bucket on absolute scale basis, i.e., a bucket with aggregated minimum = `500M`, and maximum = `25G x 8` = `200G` into 8 equal parts, i.e. of 25G each? But then this would mean that we would scale out horizontally only when each replica would reach its maximum, i.e., 25GB.
I must have misunderstood what you meant. Some examples would be helpful. :)
thanks.

Reply:
My point is, we need to clarify what we mean by linear scaling. What is being scaled linearly? What is the reference frame? Absolute scale seems to be a more reliable reference frame.

What looks linear on the vertical scale might be exponential or logarthmic depending on what is the Y axis (replicas or vertical scale).

I am not saying the linear scaling on the absolute scale is the desirable one. But the absolute scale might make a better reference frame to describe the actually desirable scaling as we have discussed in the examples already offline for apiserver.

Reply:
>  as we have discussed in the examples already offline for apiserver

Do you mean an example where we divided a bucket linearly based on absolute/aggregated resource values rather than based on resource value per replica?
I might have missed the call in which this was discussed. Can you please elaborate by providing the example here again.

Reply:
The example I meant was what we discussed based on the scaling recommendation for apiserver.  What was also mentioned in the excel from Vedran.

I am not trying re-invent that part of what way of scaling is better suited for the apiserver. I am only trying to find a way to describe the scaling for apiserver that we have already decided.

Reply:
ok. Now I am confused even further. I am not able to understand what you mean, especially if I put the statement from your first comment into the context:
>I think it makes more sense to divide up the absolute scale among the buckets rather than only the vertical scale.

Can you please try to explain from scratch. It might help if you can explain with an example too with the same bucket that is mentioned in the lines to which you gave comments.
Are you suggesting another way to divide this bucket, or just another way to describe the same "linear division" that is already described in the doc?
In case it is former, can you please provide example how that can be done. In case it the latter, can you please provide how else to describe it.
Thanks.

Reply:
All I am trying to point out is that there are hidden assumptions in the proposal in the usage of term _linear scaling_. Can you please state what is being scaled linearly explicitly? I am guessing it is the vertical scale of a single replica.

Also, could you address the question of what else could be considered for scaling (linearly or otherwise) and why what you have chosen makes more sense to use. From the document I get the feeling these assumptions haven't been explored. But that might just be due to the fact the it has not been explicitly mentioned in the document.

Reply:
I have already mentioned what I am looking for is for a way to describe the scaling behaviour. Not to change the already discussed scaling behaviour for apiserver. But that does not make it a merely academic question. The way we describe it will have an implication on how we code.

Reply:
Ah ok. will add these details.
and just to be sure, I can ignore the following statement from the first comment:
>I think it makes more sense to divide up the absolute scale among the buckets rather than only the vertical scale.

Please let me know if it is still the case, and we need to switch the proposal according to this.

Reply:
> and just to be sure, I can ignore the following statement from the first comment:

Yes. Provided if you explain why ;-)

Reply:
added in https://github.com/gardener/hvpa-controller/pull/71/commits/bab120b092e43aadf936dac3ac67a63106853e02